File: variable_client/clojure/src/opcua/variable_client.clj
(ns opcua.variable-client
  (:gen-class)
  (:require [clojure.tools.logging :refer (info)])
  (:import org.eclipse.milo.opcua.sdk.client.OpcUaClient
           org.eclipse.milo.opcua.stack.core.Identifiers
           org.eclipse.milo.opcua.stack.core.types.builtin.NodeId))

(def connection (OpcUaClient/create
             "opc.tcp://localhost:4840"))

(defn read-time
  "Read server status date time"
  []
  (let [_ (.get (.connect connection))
        address-space (.getAddressSpace connection)
        node (.getVariableNode address-space Identifiers/Server_ServerStatus_StartTime)
        value (.readValue node)]
    (info (str "Server date is: "(.getJavaDate (.getValue (.getValue value)))))))

(defn read-answer
  "Read the answer variable"
  []
  (let [_ (.get (.connect connection))
        address-space (.getAddressSpace connection)
        node (.getVariableNode address-space
                               (NodeId. 1  "answer"))
        value (.readValue node)]
    (info (str "The answer is: " (.getValue (.getValue value))))))

(defn -main
  [& args]
  (read-time)
  (read-answer)
  (.disconnect connection))

File: variable_client/cpp/variable-client.cpp
#include <format>
#include "open62541pp/open62541pp.h"
#include "open62541pp/Logger.h"

int main() {
    opcua::Client client;
    client.connect("opc.tcp://localhost:4840");

    opcua::Node node = client.getNode(opcua::VariableId::Server_ServerStatus_CurrentTime);
    const auto dt = node.readValueScalar<opcua::DateTime>();

    opcua::log(client, opcua::LogLevel::Info, opcua::LogCategory::Client,
               dt.format("Server date is: %Y-%m-%d %H:%M:%S"));

    opcua::Node answer_node = client.getNode({1, "answer"});
    const auto answer = answer_node.readValueScalar<int32_t>();
    opcua::log(client, opcua::LogLevel::Info, opcua::LogCategory::Client,
               std::format("The answer is: {}", answer));

    client.disconnect();
}
